proof/refine/ARM/Finalise_R.thy

```diff
@@ -4362,62 +4489,388 @@ global_interpretation delete_one
+  apply (rule corres_stateAssert_implied[where P'=\<top>, simplified])
+   apply (rule corres_guard_imp)
+     apply (rule corres_split_new[OF get_sc_corres, where R'="\<lambda>sc. ko_at' sc scPtr"])
+       apply (rule_tac Q'="ko_at' sc scPtr
+                and K (scReply sc = hd_opt (sc_replies x))
+                and (\<lambda>s. scReply sc \<noteq> None \<longrightarrow> reply_at' (the (scReply sc)) s)
+                and (\<lambda>s. heap_ls (replyPrevs_of s) (scReply sc) (sc_replies x))"
+              and Q="sc_at scPtr
+                and pspace_aligned and pspace_distinct and valid_objs
+                and (\<lambda>s. \<exists>n. ko_at (Structures_A.SchedContext x n) scPtr s)"
+              in stronger_corres_guard_imp)
+         apply (rule corres_guard_imp)
+           apply (rule_tac F="(sc_replies x \<noteq> []) = (\<exists>y. scReply sc = Some y)" in corres_gen_asm2)
+           apply (rule corres_when)
+            apply clarsimp
+           apply (rule_tac F="scReply sc = Some (hd (sc_replies x))" in corres_gen_asm2)
+           apply clarsimp
+           apply (rule corres_split_new[OF replyNext_update_corres_empty])
+             apply (rule update_sc_reply_stack_update_ko_at'_corres)
+            apply wpsimp+
+          apply (clarsimp simp: obj_at_def)
+          apply (frule (1) valid_sched_context_objsI)
+          apply (clarsimp simp: valid_sched_context_def list_all_def obj_at_def)
+         apply clarsimp
+         apply (case_tac "sc_replies x"; simp)
+        apply assumption
+       apply (clarsimp simp: obj_at_def  is_sc_obj)
+       apply (frule state_relation_sc_replies_relation)
+       apply (subgoal_tac "scReply sc = hd_opt (sc_replies sca)")
+        apply (intro conjI)
+          apply clarsimp
+         apply clarsimp
+         apply (erule (1) reply_at_cross[rotated])
+          apply (frule (1) valid_sched_context_objsI)
+          apply (clarsimp simp: valid_sched_context_def list_all_def obj_at_def)
+         apply fastforce
+        apply (erule (1) sc_replies_relation_prevs_list)
+        apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def projectKO_sc)
+       apply (erule (1) sc_replies_relation_scReply[symmetric])
+       apply (clarsimp simp: obj_at'_real_def  ko_wp_at'_def projectKO_sc)
+      apply wpsimp+
+    apply (fastforce simp: sym_refs_asrt_def)+
+  done
+
+lemma schedContextUnbindYieldFrom_corres:
+  "corres dc (einvs and sc_at scPtr and K (scPtr \<noteq> idle_sc_ptr)) (invs' and sc_at' scPtr)
+          (sched_context_unbind_yield_from scPtr) (schedContextUnbindYieldFrom scPtr)"
+  apply (clarsimp simp: sched_context_unbind_yield_from_def schedContextUnbindYieldFrom_def
+                        maybeM_when)
+  apply add_sym_refs
+  apply (rule corres_stateAssert_implied[where P'=\<top>, simplified])
+   apply (rule corres_guard_imp)
+     apply (rule corres_split_new[OF get_sc_corres])
+       apply (case_tac sca)
+       apply (clarsimp simp: sc_relation_def)
```

Comment:
rename_ac this sca and is the indentation correct here?

Reply:
Yep, the indentation is correct. The `case_tac` doesn't introduce any new cases, it's instead being used to expand the `SchedContext` datatype constructor so that the following `wp_fix` can be used.